### PR TITLE
Login popup window

### PR DESCRIPTION
**Pull request by JayWelborn**
*Description of changes:*

This PR addresses comments on #192 and adds options to position the Login Popup window. #192 is a good improvement to the login flow that makes the CCP experience match other 3rd-party login flows, and I think it makes a good addition to the Project. 

Since that PR hasn't had any updates in a few months, I'm opening a new PR with the requested changes to try and get this feature into the main library.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/core.js
Comment (jolim24601) [inline src/core.js:527]: This will cause a null exception if loginPopup is `true`, you need to do a `instanceof` check here.

**Review by jolim24601 — CHANGES_REQUESTED** [review]
Please verify there is no null exception and that there is backwards compatibility here

File: src/core.js
Comment (JayWelborn) [reply src/core.js:527]: If `params.loginPopup` is `true`, `params.loginPopup.forceWindow` is undefined, with no null exception. (You can test in FF/Chrome browser console with `true.testProp` or `true.testProp === undefined`)

The `params.loginPopup && params.loginPopup.forceWindow` will return false because `true && undefined` is `false`, so options will be `{}`.

File: Documentation.md
Comment (jolim24601) [inline Documentation.md:143]: Could you please also update the `README.md` file as well? Also, please run `npm run release` to update the build files that are in our `release` folder. We will automate this in the future but for now it's a good idea to run this every time a change is made so that they're not out of sync.

**Review by jolim24601 — CHANGES_REQUESTED** [review]
Update README and update release build

File: Documentation.md
Comment (JayWelborn) [reply Documentation.md:143]: Updated `README.md` to match the changes in `Documentation.md` and rebuilt the release files.

**Review by jolim24601 — APPROVED** [review]

**Comment by jolim24601** [comment]
Pinged a teammate to review as we need 2 approvals to merge.

**Comment by JayWelborn** [comment]
> Pinged a teammate to review as we need 2 approvals to merge.

Thanks @jolim24601 . I can't request reviewers, so I'm not exactly sure what the steps are from here.

**Review by h3adache — CHANGES_REQUESTED** [review]
After discussion internally. it would be preferred to keep the old booleans unchanged and add a new options object similar to `softphone`.
Sorry this is a change from the discussion on #192 
Can you please separate these?

**Comment by JayWelborn** [comment]
> After discussion internally. it would be preferred to keep the old booleans unchanged and add a new options object similar to `softphone`.
> Sorry this is a change from the discussion on #192
> Can you please separate these?

Sure thing. If both a `loginPopup` boolean and a `loginWindow` options object are supplied, should the options object take precedence over the bool?

**Comment by h3adache** [comment]
I think we can keep it simple and check the old boolean values first followed by the new setting.
`loginPopup || loginOptions.usePopup`
I changed the names based on my feedback about the naming but feel free to take them or suggest other ones.

File: Documentation.md
Comment (JayWelborn) [reply Documentation.md:166]: Restored documentation for autoClose, and reverted to original docs for loginPopup since it's back to being a boolean.

File: src/util.js
Comment (JayWelborn) [reply src/util.js]: I moved these to the top of the file with other constants.

These were from the old PR, but after some experimenting 578px tall is about where the window gets a vertical scrollbar (576/577 depending on which debugger I was using). That width has about the same amount of space on the sides as above/below the login widget in the popup window.

File: src/util.js
Comment (JayWelborn) [reply src/util.js]: After looking at a couple ways to do it, I went with

```js
var top = isNaN(options.top) ? 0 : options.top
```

to avoid having to build the window features string on the fly, and still have some fallback behavior if the caller gives bad/no value. I think letting users center it themselves is reasonable.

**Comment by JayWelborn** [comment]
@h3adache the update commit restores the boolean behavior of `loginPopup` and `loginPopupAutoClose`. It adds an optional `loginOptions` param that will be ignored if `loginPopup` is set to `false`.

From a library user's point of view it makes sense to me that setting `loginPopup` to `false` wouldn't open a popup even if I gave an object full of options.

**Comment by h3adache** [comment]
Yea that makes sense to me! We can always push that option to the new options param in future. That's a nice perk of having that separate object 😃

File: src/util.js
Comment (h3adache) [inline src/util.js:422]: Not sure if you are done yet but the 'canonical' way to do these (there are examples in other areas in this project as well) is by using javascript's short cut as the elvis operator.
For example
```
const options = {}
var popupTop = options.top || 0
console.log(popupTop)
options.top = 10
popupTop = options.top || 0
console.log(popupTop)
```

File: src/util.js
Comment (h3adache) [inline src/util.js:436]: extra space here

File: src/core.js
Comment (h3adache) [inline src/core.js:527]: Because of the way you are checking this below you don't need to do this check here.
It's just assigning `false` to something already `false` if they don't supply it.
You can just pass the `params.loginOptions` directly to `open`

File: src/util.js
Comment (h3adache) [reply src/util.js]: Sorry I made a comment on this before reading your explanation here.
I think it's very good of you to check but since you have clear documentation already here, we can just leave it to the clients to work out bad inputs :)

File: src/util.js
Comment (JayWelborn) [reply src/util.js]: updated to

```js
var height = options.height || DEFAULT_POPUP_HEIGHT;
var width = options.width || DEFAULT_POPUP_WIDTH;
var top = options.top || 0;
var left = options.left || 0;
```

File: src/util.js
Comment (JayWelborn) [reply src/util.js:436]: Before, this was indented one space less than `connect.PopupManager.prototype.open` above.

File: src/core.js
Comment (JayWelborn) [reply src/core.js]: Updated to remove unneeded safety check

**Review by ctwomblyamzn — APPROVED** [review]

**Review by h3adache — APPROVED** [review]

**Comment by h3adache** [comment]
We have a squash and merge procedure that didn't work cleanly with this but your changes are merged.
Thanks you for the contribution @JayWelborn!

**Comment by tomsabin** [comment]
Is there a plan to release this to NPM? The current release of v1.5.0 does not include this unfortunately.

**Comment by tomsabin** [comment]
@h3adache Is there a plan to release this to NPM? The current release of v1.5.0 does not include this unfortunately.

**Comment by h3adache** [comment]
Hi @tomsabin thanks for the reminder!
Version 1.5.1 is released to NPM now and includes this feature

**Comment by nickgal** [comment]
This no longer functions with 1.6.3 / 1.6.4 and appears to have been reverted with https://github.com/amazon-connect/amazon-connect-streams/pull/360 as
mentioned by https://github.com/amazon-connect/amazon-connect-streams/issues/362 and https://github.com/amazon-connect/amazon-connect-streams/pull/384

**Pull request closed without merging**
